lib/core/scheduler.cpp

```diff
@@ -0,0 +1,39 @@
+#include "scheduler.hpp"
+#include "wall_clock.hpp"
+
+namespace hyped::core {
+
+Scheduler::Scheduler(core::ILogger &logger, core::ITimeSource &time)
+    : logger_(logger),
+      time_(time),
+      task_queue_()
+{
+}
+
+core::Result Scheduler::run()
+{
+  if (task_queue_.empty()) { return core::Result::kSuccess; }
+  const auto current_time = std::chrono::time_point_cast<std::chrono::nanoseconds>(time_.now());
+  auto next_task          = task_queue_.top();
+  core::Result result     = core::Result::kSuccess;
+  for (std::size_t i = 0; i < task_queue_.size(); i++) {
+    if (current_time < next_task.sheduled_time) { return result; }
+    core::Result next_result = next_task.handler();
+    if (next_result == core::Result::kFailure) { result = core::Result::kFailure; }
+    task_queue_.pop();
+    if (task_queue_.empty()) { break; }
+    next_task = task_queue_.top();
+  }
+  return result;
+}
```

Comment:
The logic flow here seems a little odd to me. I understand why one would write it like this but I am unsure that that's what we want.

I think the logic for something like this should be a simple and clean as possible:
```
while next task is due do
  run next task
  if task failed, return failure
return success
```

I understand that there's a bit more bookkeeping to do. So it would be closer to
```
get current time
get number of tasks scheduled
for at most the number of task scheduled iterations
   get next task
   if task is not due, return success
   run task
   if task failed, return failure
return success
```

Here are some highlights why I think this is better than what you have so far:
1. The call to `.size()` happens before the loop so we avoid any weird behaviour where we take out elements while incrementing `i`, essentially incrementing `i` twice. (i.e. if all tasks are due, then you may only end up running half of them)
2. We removed the extra case of no tasks being due. The extra call to get the current time doesn't seem to be a big deal, and it certainly won't be an issue when the system is under load (because then the queue won't be empty anyways).
3. We avoid the mutable variables `next_task` and `result`
4. The logic has changed a little: I understand a task failing to be an irrecoverable failure on a system level. Maybe we should think about building in some shut down handlers in this case, but I wouldn't expect other tasks to be run after such a failure. Any internal error handling should be done by the tasks through other channels.